[RDY] Allow functions called by statusline %! option to access the statusline-local window handle

joshuagiles opened the pull request:
The two options for statuslines which evaluate expressions, %{} and %! have different behavior regarding which buffer they work on. The documentation states that %{} operates in the statusline-local context, so its 'current buffer' is whichever one the statusline is being evaluated for. However, when the %{} expression is evaluated, a temporary variable g:actual_curbuf is set, to allow the expression to access either the statusline's buffer, or the actual global current buffer. %! has no such option.

This means that if a user wants something in the %! expression to change behavior based on the buffer of its calling statusline, they must output %{} blocks which are then evaluated after the %! is finished. But of course, %{} expressions do not produce evaluated output themselves, only raw text. This means that it is regular for users to have to construct awkward and unintuitive systems in the %! function to output the correct %{} expressions.

For example, if I wanted for my statusline to have a different color for one section based on a buffer-local variable, it would require something like this (a much more simple example than what some advanced users might want from this feature):

```
function! MyStatusLine()
    let line =  '%1*%{b:foobar?"My Statusline":""}'
    let line .= '%2*%{!b:foobar?"My Statusline":""}'
    return line
endfunction

let statusline=%!MyStatusLine()
```

While I'm sure many experienced vim users are accustomed to this style of statusline function magic, for a person who is making a statusline for their first time, and expecting the %! expression to allow for more-or-less straightforward programmatic statusline formatting, it is highly unintuitive.

However, if the function called from %! has access to the statusline's local window handle, it can encapsulate all of this logic without having to output %{} blocks to be evaluated afterward. This allows much more intuitive statusline function programming. Using the same example as above:

```
function! MyStatusLine()
    let curbuf = nvim_win_get_buf(g:stl_window)
    let section_color = nvim_buf_get_var(curbuf, 'foobar') ? '%1*' : '%2*'
    return section_color . 'My Statusline'
endfunction
```

This is a very small code change, using nearly identical code around the %! evaluation, as is around the %{} evaluation. Just setting a variable before the call to eval_to_string_safe, and then unsetting it afterward. However, the usability improvement is quite large just going by my own statusline tweaking work which I have done while using this feature.

joshuagiles committed:
Add set_internal_number_var function in eval.c

A counterpart to set_internal_string_var

joshuagiles committed:
Change g:actual_curbuf variable to use set_internal_number_var

Used when evaluating %{} blocks in the statusline, the variable g:actual_curbuf
is set so the expressions can see both the statusline-local current buffer, as
well as the global, actual current buffer. However, because it used
set_internal_string_var, actually using this variable in vimscript required
coercing it to a number variable before it could be used in any API calls, to
avoid an "Expected Buffer" error. Now the g:actual_curbuf variable can be used
directly.

tjdevries commented:
Maybe consider changing this to a `v:` variable, since it's built-in and should not be edited by the user. Don't know the specifics on how to do that though.

joshuagiles commented:
That's a good suggestion. I made the new stl_window `g:` because actual_curbuf was so originally (even though it wasn't set correctly), but I agree it makes more sense this way.

tjdevries commented:
So, we can't just throw away the old "g:actual_curbuf" because that would break statusline implementations (I forgot about the variable name. I thought it was a new one that you were adding).

I'd add documentation saying that `g:actual_curbuf` is deprecated, but don't remove it. Leave that as it was before. Add some new internal `v:` variables that solve the problems you're experiencing.

Also, you should add some tests. All PRs need tests. :)

justinmk commented on src/nvim/eval.c:
```diff
@@ -400,6 +400,8 @@ static struct vimvar {
   VV(VV_HLSEARCH,       "hlsearch",         VAR_NUMBER, 0),
   VV(VV_OLDFILES,       "oldfiles",         VAR_LIST, 0),
+  VV(VV_ACTUAL_CURBUF,  "actual_curbuf",    VAR_NUMBER, VV_RO),
+  VV(VV_STL_WINDOW,     "stl_window",       VAR_NUMBER, VV_RO),
```
Probably `v:actual_curwin` is a better choice. It's not really statusline-specific, semantically, though that happens to be the only place it's useful. And even if existing names are ugly, it's better to follow their pattern than add a new pattern.

Also agreed with @tjdevries : we need to keep `g:actual_curbuf`, but in the documentation move it to `deprecated.txt`. And introduce a redundant `v:` variant, `v:actual_curbuf`.

bfredl replied:
But this is not at all `curwin`, rather "the window for which the statusline is currently drawn". It is the logical opposite of `actual_curbuf`.

justinmk replied:
Ah, I see. Hmm.

bfredl replied:
If you mean new var, why not `v:stl_buffer` ? If you mean `v:actual_curbuf` I see no point of renaming it.

justinmk replied:
What i mean is if the two variables `v:actual_curbuf` and `v:stl_window` are analogous they should be named analogously.

justinmk replied:
Alright, I should not have commented without looking closely at this stuff. This is extremely confusing.

joshuagiles committed:
Document changes to statusline window/buffer variables

joshuagiles commented:
I'd love to add tests but I don't know lua at all, and it looks like adding tests for this feature would require adding tests all the way back up to %! and %{} statusline options themselves, which are not covered in any of the current tests (nor the old variable g:actual_curbuf, see below). It also would require further work in the statusline tests to allow calling a vimscript function which uses the relevant variables from a statusline, which to be honest is outside what I think I'm personally capable of, given that I don't really know how the tests are structured or written here to begin with.

Of course it's a problem that there is no test coverage for nearly any of the relevant parts of the code, pre-extant or added in this PR, but I hope anyone can still see the value of having this information exposed to folks trying to write a statusline using the %! option, as it is when using the %{} option.

The g:actual_curbuf variable was partially broken before I started working on this; if anyone used a function referring to that variable, while also using a plugin or other user-defined function which used the redraw command, the variable would not be found, because it was set as only "actual_curbuf" (while being unset as "g:actual_curbuf"). This PR also fixes this issue.

Anyway, I've added documentation for the new variable, and the modified v:actual_curbuf variable, and put a note in deprecated.txt about the scope change for the latter. Even if this PR doesn't merged, I at least hope someone will recognize the benefit in the core idea here.

KillTheMule commented:
Just in case, see https://github.com/neovim/neovim/pull/7882.

bfredl commented:
I don't think we should follow #7882 for our own status line tests. Rather use screen test with `set laststatus=2`

KillTheMule commented:
Yeah, I just wanted to point out that there are now tests for the statusline, in case it went unnoticed.

joshuagiles committed:
Add statusline window variables v:stl_winid and v:actual_curwinid

`v:stl_winid` is set during evaluation of `%!` statuslines, and holds the
ID of the window the statusline is currently being drawn for.

`v:actual_curwinid` is set during evaluation of `%{}` statusline expressions,
and holds the ID of the actual current window (the one the user is editing in),
rather than the window the statusline is currently being drawn for.

This allows `%!` and `%{}` in statuslines to have rough parity in their
capabilities and makes certain statusline highlighting situations much easier
to configure.

joshuagiles commented:
To remedy some confusion, here's a little better explanation of what's going on here, with the original behavior, and what I've done to improve it.

The two statusline options `%!` and `%{}` have similar but subtly different behavior:

* `%{}` operates in the context of the window/buffer the statusline is being evaluated for.
    * Because of this, the variable `actual_curbuf` was provided (before this PR) to access the 'actual' global current buffer. The regular functions to get the current buffer won't work because nvim actually sets those variables temporarily around evaluation of this expression, so `nvim_get_current_buffer()` will not return the buffer that's actually being edited.
    * Its output is *not* re-evaluated for nested %{} expressions, nor references to variables or highlight groups.
* `%!` operates in the context of the global current window/buffer (opposite to how `%{}` works)
    * No special variable is provided to allow a function called this way to know what the statusline-local buffer is.
    * Its output *is* re-evaluated for nested %{} expressions, or references to variables or highlight groups

So, one can see that the two options do not have feature parity. One's output is re-evaluated, but it doesn't have access to the same breadth of information as the other. This has led to a generally-accepted method of writing statuslines by making a `%!` function which, because it doesn't have access to the same information as `%{}`, has to nimbly perform whatever logic it's capable of, but then output what it isn't capable of in `%{}` blocks, which have their own limitation of not having their results re-evaluated, and very limited information-sharing between the two options. See the two examples in the initial PR text above.

The idea here is to have a variable which is set around `%!` evaluation, just like one is set around `%{}` evaluation, so that both have access to the same information, allowing a user to write a cleaner, easier-to-understand function which is able to do all the logic it needs to right in the `%!` function, instead of having to come up with clever ways to pass certain patterns forward to `%{}` expressions.

I think some confusion may have arisen from the fact that one refers to a window, while the other refers to a buffer. This was a matter of usefulness: if a statusline only knows what buffer it's being evaluated for but not necessarily the window, it would be identical for two windows on the same buffer, even if one window is the active one, while the other is not. In some discussions about this in the IRC channel, several people noted hat statuslines are *window* local, rather than *buffer* local. This general community understanding, alongside the fact that this actual_curbuf feature seems to have been not well-known or widely-used enough for anyone to have noticed its associated bugs, is why I chose to provide the current *window* for the new feature I'm adding here. In order to fix this bit of confusing naming, and based on a discussion in IRC with justinmk, I've decided to make things a bit more intuitive by changing the two variables to `v:stl_winid` and `v:actual_curwinid`, both now referring to a window rather than a buffer.

joshuagiles commented:
Here are some simple examples of things a person might do with this feature:

Special statusline highlight for buffers with filetype 'vim':
```
function! MyStatusLine()
    let stlbuf = nvim_win_get_buf(v:stl_winid)
    let buf_ft = nvim_buf_get_option(stlbuf, "filetype")
    let stlhighlight = (buf_ft == 'vim') ? '%1*' : ''
    return stlhighlight . '%f'
endfunction
```

Extra statusline information for minimized windows:
```
function! MyStatusLine()
    let stlbuf = nvim_win_get_buf(v:stl_winid)
    let winheight = nvim_win_get_height(v:stl_winid)
    if winheight == 0
        return '%f + Extra information'
    else
        return '%f'
    endif
endfunction
```

Different statusline information for inactive windows:
```
function! MyStatusLine()
    if nvim_get_current_win() != v:stl_winid
        return '%f + Inactive window'
    else
        return '%f'
    endif
endfunction
```

janlazo commented:
Vim patch 8.1.1372, https://github.com/neovim/neovim/pull/12538/commits/edaff441de1015068d7cee64c222f3c65fcd7dc6, was ported in https://github.com/neovim/neovim/pull/12538/commits/edaff441de1015068d7cee64c222f3c65fcd7dc6 to set `g:statusline_winid` when evaluating `%!`.

Does that cover the use-case, solved by this PR?

tjdevries commented:
Yeah, I think so. Let's close this. Thanks @janlazo
